### PR TITLE
Encapsulate HPPC usage inside of ImmutableOpen(Int)Map

### PR DESCRIPTION
**Pull request by dakrone**
This commit encapsulates a lot of the usage of HPPC into `ImmutableOpenMap` and `ImmutableOpenIntMap`.

To accomplish this, this commit adds the `entrySet()` method to these classes returning `Set<Map.Entry<KType, VType>>`. This allows a map to be streamable using java Streams (`myMap.entrySet().stream()`) as well as ensuring that using a Java `foreach` loop **does not leak the underlying HPPC implementation outside of the ImmutableOpenMap**. This means that a variable could be switched from `ImmutableOpenMap` to `HashMap` without affecting the traversal of its contents.

The rest of this PR is migrating usages of the HPPC iterators (`IntObjectCursor`, `ObjectCursor`, `ObjectObjectCursor`) to use `entrySet()` and regular java `Map.Entry<>` looping.

I also ran benchmarks with Rally before and after these changes, and wasn't able to detect an appreciable difference between these.

Subsequent work for this would involve migrating the rest of the Java source off of HPPC, and then experimenting with moving our collection use from `ImmutableObjectMap` to something like `HashMap`, or potentially backing `ImmutableObjectMap` with a built-in Java collection if we decided we want to keep it around. This PR was getting big enough that I felt it useful to open it as progress and for discussion rather than trying to do everything at once.

### CODE REVIEW

**Comment by joegallo** [comment]
The code I see here looks good -- great change in the callers to see it feel more 'normal'.

I was surprised to see, for example, that there are still more than just a couple of usages of `ObjectObjectCursor` remaining:

```
joegallo@galactic:~/Code/elastic/elasticsearch $ git grep 'import com.carrotsearch.hppc' | grep -v /test/ | grep -v /internalClusterTest/ | grep Cursor | grep -v common/collect | grep ObjectObject
client/rest-high-level/src/main/java/org/elasticsearch/client/ilm/GetLifecyclePolicyResponse.java:import com.carrotsearch.hppc.cursors.ObjectObjectCursor;
server/src/main/java/org/elasticsearch/search/dfs/AggregatedDfs.java:import com.carrotsearch.hppc.cursors.ObjectObjectCursor;
server/src/main/java/org/elasticsearch/search/dfs/DfsSearchResult.java:import com.carrotsearch.hppc.cursors.ObjectObjectCursor;
x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/cluster/metadata/MetadataMigrateToDataTiersRoutingService.java:import com.carrotsearch.hppc.cursors.ObjectObjectCursor;
x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlInitializationService.java:import com.carrotsearch.hppc.cursors.ObjectObjectCursor;
x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/MappingsMerger.java:import com.carrotsearch.hppc.cursors.ObjectObjectCursor;
x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/extractor/ExtractedFieldsDetectorFactory.java:import com.carrotsearch.hppc.cursors.ObjectObjectCursor;
x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobDataDeleter.java:import com.carrotsearch.hppc.cursors.ObjectObjectCursor;
x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/local/LocalExporter.java:import com.carrotsearch.hppc.cursors.ObjectObjectCursor;
x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/index/IndexResolver.java:import com.carrotsearch.hppc.cursors.ObjectObjectCursor;
x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/action/TransportGetRollupCapsAction.java:import com.carrotsearch.hppc.cursors.ObjectObjectCursor;
```

Loooking closer at `TransportGetRollupCapsAction` (it's the last one, so it was closest to my cursor), it seems like the same migration you've used elsewhere would also work fine there.

What's the dividing line between the changes you did make on this PR versus the others that are still as yet undone? 

Note: something like "Oh, I never intended to personally complete all of these myself on this one PR, this is just meant to show the way and others can migrate as they'd like" is almost a perfectly acceptable answer. The "almost" is that I would kindof like to actually see the thing through at some point, so I guess I'm curious what the plan is to see this through to completion.

**Comment by dakrone** [comment]
@joegallo that's right, there are a lot of uses of bothe `ObjectObjectCursor` and `ObjectCursor` remaining. The dividing line for this PR is mostly size (I thought it was getting on the "large" side, and could easily be split into multiple pieces of work) as well as timing. Since this was a spacetime project, I thought it would be best to open this to have concrete progress made, with the caveat that we can continue to make progress on it in an ad-hoc manner once this has been merged.

That, and the larger this PR gets, the more difficult it will be to backport and the higher the chance of merge conflicts due to inadvertent changes.

> so I guess I'm curious what the plan is to see this through to completion.

If this PR gets approved, I'll open a public issue to track making the rest of the migrations. It will remain a fairly low priority, however, so the rest of the migrations may be done "on-demand". There remain a few tricky ones that I haven't tackled in this PR either.

File: server/src/main/java/org/elasticsearch/common/collect/ImmutableOpenMap.java [hunk 1]
```diff
@@ -112,6 +118,103 @@ public boolean isEmpty() {
...
+        }
+
+        @SuppressWarnings("unchecked")
+        public boolean contains(Object o) {
+            if (o instanceof Map.Entry<?, ?> == false) {
+                return false;
+            }
+            Map.Entry<?, ?> e = (Map.Entry<?, ?>) o;
+            Object key = e.getKey();
+            return map.containsKey((KType) key);
```
Comment (joegallo) [inline server/src/main/java/org/elasticsearch/common/collect/ImmutableOpenMap.java:199]: I don't think only checking the key here is sufficient.

File: server/src/main/java/org/elasticsearch/common/collect/ImmutableOpenIntMap.java [hunk 2]
```diff
@@ -140,6 +150,102 @@ public void remove() {
...
+            return new ConversionIterator();
+        }
+
+        public boolean contains(Object o) {
+            if (o instanceof Map.Entry<?, ?> == false) {
+                return false;
+            }
+            Map.Entry<?, ?> e = (Map.Entry<?, ?>) o;
+            Object key = e.getKey();
+            return map.containsKey((Integer) key);
```
Comment (joegallo) [inline server/src/main/java/org/elasticsearch/common/collect/ImmutableOpenIntMap.java:230]: Same.

**Review by joegallo — CHANGES_REQUESTED** [review]
The `.contains(Object o)` for the entrySet should, I think, be checking both keys and **values**, at least, that's how the HashMap version of same appears to behave to me.

**Review by joegallo — APPROVED** [review]
🚢

**Pull request merged**